### PR TITLE
Create VirtualizedGrid component for catalog view

### PR DESCRIPTION
**Pull request by sahil143**
**Fixes**: https://issues.redhat.com/browse/ODC-4027

**Analysis**:  Virtualize Developer Catalog to support very large number of items

**Solution**: Created a VirtualizedGrid component which is used in catalog
                      **NOTE: this also affects the operator hub because same component is shared between the catalog page and operator hub**

**Screenshots**:

GroupBy-min

**Unit tests**:
Screenshot from 2020-06-22 22-16-48
Screenshot from 2020-06-23 13-33-45

Browser:

- [x] Chrome

### CODE REVIEW

**Comment by rohitkrai03** [comment]
@sahil143 I haven't gone into reviewing code in depth yet, but noticed a few issues by just running your branch locally. To start off with the performance really seems to be slow and the frame rates are dropping too much with the new implementation. While scrolling the screen feels janky due to very low frame rates. 

Below is a gif of dev catalog on your branch, notice the low frame rate shown in the top left corner. Its almost always below 10fps.
Peek 2020-06-24 02-12

Below is a dev catalog on master, notice how the frame rate goes up to 60fps.

Peek 2020-06-24 02-19

**Comment by sahil143** [comment]
> I also see a lot of react warnings in the console.

fixed, except for the key prop missing. Key Prop is provided on both CellMeasurer and wrapper div but somehow it's being removed in react-virtualized.

> I see the bottom of the tiles being cut off when there are less number of items in catalog -

height of the container is being set by the window scroller and the tiles are not cut it's just that their box shadow is hidden. i tried adding margin which adds the space but the shadow is still hidden for the last row of tiles.

> Also noticed this weird issue where the tiles suddenly vanish from the top while scrolling up and down and then come back suddenly.

I wasn't able to reproduce this while scrolling. Is there any particular step that needs to be followed here?

>To start off with the performance really seems to be slow and the frame rates are dropping too much with the new implementation. While scrolling the screen feels janky due to very low frame rates.

yes, the frame rate seems to be very low, Investigating this issue.

**Comment by christianvogt** [comment]
Scrolling is a bit janky it seems because the scrollbar grows and shrinks in size as your scroll down and up. I would expect the scrollbar to remain relatively the same size. After scrolling through the entire list, I would expect the cache to have a precise measurement on the scrollbar and to no longer adjust the sizing. But that's not the case here.

File: frontend/packages/console-shared/src/components/virtualized-grid/Cell.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  };
+  return item ? (
+    <CellMeasurer
+      cache={cache}
+      columnIndex={columnIndex}
+      key={key}
+      parent={parent}
+      rowIndex={rowIndex}
+    >
+      <div style={wrapperStyles} key={key}>
```
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/virtualized-grid/Cell.tsx:46]: remove `key`

File: frontend/packages/console-shared/src/components/virtualized-grid/GroupByFilterGrid.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  items: groupedItems,
+  children,
+}) => {
+  const idealItemWidth = IDEAL_CELL_WIDTH + IDEAL_SPACE_BW_TILES;
+  const columnCountEstimate = Math.max(1, Math.floor(width / idealItemWidth));
+  const { items, rowCount, columnCount, headerRows } = getItemsAndRowCount(
+    groupedItems,
+    columnCountEstimate,
+  );
+  const cache = new CellMeasurerCache({ defaultHeight: 250, minHeight: 200, fixedWidth: true });
```
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/virtualized-grid/GroupByFilterGrid.tsx:37]: defaults should be configurable

**Comment by christianvogt** [comment]
The tiles rendered to screen seem to be re-rendering on every scroll. I'd expect them to not have to re-render unless data changes.

File: frontend/packages/console-shared/src/components/virtualized-grid/utils.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+ */
+export const getHeightAndWidthOfCell = (
+  h: string | number,
+  w: string | number,
+  item: CellItem,
+): CellSize => {
+  let height: string | number;
+  let width: string | number;
+  if (!item) return { height, width };
+  const isItemString = typeof item === 'string';
```
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/virtualized-grid/utils.ts:102]: Is this the only way to distinguish a group header from tile?

**Comment by christianvogt** [comment]
The last row in the grid disappears prematurely when scrolling up.

**Comment by sahil143** [comment]
> Scrolling is a bit janky it seems because the scrollbar grows and shrinks in size as your scroll down and up. I would expect the scrollbar to remain relatively the same size. After scrolling through the entire list, I would expect the cache to have a precise measurement on the scrollbar and to no longer adjust the sizing. But that's not the case here.

 the scrollbar is growing and shrinking because of dynamic row/cell height, I tried making the row height fixed and the scrollbar remained the same size.

>The tiles rendered to screen seem to be re-rendering on every scroll. I'd expect them to not have to re-render unless data changes.
edit: the items are actually being unmounted and remounted on scroll

I guess that's more or less expected, this issue talks about the same https://github.com/bvaughn/react-virtualized/issues/416

>The last row in the grid disappears prematurely when scrolling up

fixed.

File: frontend/packages/console-shared/src/components/virtualized-grid/utils.ts (hunk 3 above)
Comment (sahil143) [reply frontend/packages/console-shared/src/components/virtualized-grid/utils.ts:104]: array items only contain a string for the header and an object for item. We could make each item an object having a property `type` which can have value either `header` or  'cell`. WDYT?

**Comment by christianvogt** [comment]
> > Scrolling is a bit janky it seems because the scrollbar grows and shrinks in size as your scroll down and up. I would expect the scrollbar to remain relatively the same size. After scrolling through the entire list, I would expect the cache to have a precise measurement on the scrollbar and to no longer adjust the sizing. But that's not the case here.
> 
> the scrollbar is growing and shrinking because of dynamic row/cell height, I tried making the row height fixed and the scrollbar remained the same size.

Shouldn't we be able to assume a certain height per row?
The difference in scrollbar size on page load and after scrolling is quite noticeable.

On page load:
image

After scrolling to the bottom:
image

After scrolling back to the top:
image

Why isn't the size cached? I'd expect the scrollbar to adjust a bit while we render a tile for the first time and calculate the height. But after scrolling through the entire list, the scrollbar should become a fixed size as all known heights have been computed.

**Comment by christianvogt** [comment]
> > The tiles rendered to screen seem to be re-rendering on every scroll. I'd expect them to not have to re-render unless data changes.
> > edit: the items are actually being unmounted and remounted on scroll
> 
> I guess that's more or less expected, this issue talks about the same [bvaughn/react-virtualized#416](https://github.com/bvaughn/react-virtualized/issues/416)

This is not expected. Components should be unmounted and remounted on scroll.
Furthermore, tiles can be wrapped with `memo` to ensure we don't need to re-render.

**Comment by christianvogt** [comment]
Missing padding at the bottom. Also it seems that the background is now focusable which causes a focus outline (see black outline). Previously it was not.

image

**Comment by sahil143** [comment]
> Shouldn't we be able to assume a certain height per row?
The difference in scrollbar size on page load and after scrolling is quite noticeable.

yes, I missed a prop `estimatedRowHeight`. After adding this prop scroll bar size is consistent.

>This is not expected. Components should be unmounted and remounted on scroll.
Furthermore, tiles can be wrapped with memo to ensure we don't need to re-render.

As discussed, Added `React.memo` to the renderTile component to prevent unnecessary re-renders.

> Missing padding at the bottom.

fixed this. used padding for each cell instead of subtracting a constant from height and width to create space between tiles

Updated Screenshots
Screenshot 2020-07-08 at 12 05 32 AM
Screenshot 2020-07-08 at 12 05 56 AM

File: frontend/packages/console-shared/src/components/virtualized-grid/Grid.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    CellMeasurementContext,
+  );
+  const itemCount = items.length;
+  const idealItemWidth = cellWidth + cellMargin;
+  const columnCountEstimate = Math.max(1, Math.floor(width / idealItemWidth));
+  const rowCount = Math.ceil(itemCount / columnCountEstimate);
+  const columnCount = Math.max(1, itemCount && Math.ceil(itemCount / rowCount));
+  const cellRenderer = (data: GridCellProps) => children({ data, columnCount, items, rowCount });
+  return (
+    <GridComponent
```
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/virtualized-grid/Grid.tsx:26]: Remove tab index from Grid:
```suggestion
    <GridComponent
        tabIndex={null}
```

File: frontend/packages/console-shared/src/components/virtualized-grid/Grid.scss [hunk 5]
```diff
@@ -0,0 +1,6 @@
+.ocs-grid {
+  .ReactVirtualized__Grid__innerScrollContainer {
+    width: 100% !important;
+    max-width: 100% !important;
```
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/virtualized-grid/Grid.scss:4]: Why is this needed? I removed this and it worked.

File: frontend/packages/console-shared/src/components/virtualized-grid/VirtualizedGrid.tsx [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+              <AutoSizer disableHeight>
+                {({ width }: Size) => (
+                  <div ref={registerChild}>
+                    {isItemsGrouped ? (
+                      <GroupByFilterGrid {...props} width={width} items={items as GroupedItems}>
+                        {(gridProps: GridChildrenProps) => (
+                          <Cell
+                            {...gridProps}
+                            renderHeader={renderHeader}
+                            renderCell={renderCell}
```
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/virtualized-grid/VirtualizedGrid.tsx:72]: I wonder if this is where the lack of a `key` warning is coming from?

File: frontend/packages/console-shared/src/components/virtualized-grid/Grid.scss (hunk 5 above)
Comment (sahil143) [reply frontend/packages/console-shared/src/components/virtualized-grid/Grid.scss]: This was added to give inner container full width, but it's not necessary anymore. Removed it.

File: frontend/packages/console-shared/src/components/virtualized-grid/VirtualizedGrid.tsx (hunk 6 above)
Comment (sahil143) [reply frontend/packages/console-shared/src/components/virtualized-grid/VirtualizedGrid.tsx]: yes this was the cause for key warning in console.

**Comment by sahil143** [comment]
> I see this console:
> image
> image
> 
> When using the type filters, even when there's enough space for more columns, sometimes it shows less:
> image

@christianvogt fixed both the issue.

updated screenshot
Screenshot 2020-07-09 at 6 18 55 PM

**Comment by rohitkrai03** [comment]
@sahil143 It would also be good if you could squash your commits into better explanatory commit messages around the feature.
